crates/proto/src/dnssec/dnssec_dns_handle/mod.rs

```diff
@@ -41,17 +42,17 @@ use nsec3_validation::verify_nsec3;
 /// This wraps a DnsHandle, changing the implementation `send()` to validate all
 ///  message responses for Query operations. Update operation responses are not validated by
 ///  this process.
-#[derive(Clone)]
 #[must_use = "queries can only be sent through a DnsHandle"]
-pub struct DnssecDnsHandle<H> {
+pub struct DnssecDnsHandle<H, T = TokioTime> {
```

Comment:
Instead, could we give `DnsHandle` an associated type for the `RuntimeProvider`?

Reply:
I'm wondering why this is preferred since the handle isn't relying on anything else from the `RuntimeProvider`. Reviewing the latest push to this branch it felt to me like parameterizing a few of the types with the `Time` trait instead would be clearer.

Is the working theory that in the future these types may need more generic params and so to avoid having a bunch of different generics added we should take the `RuntimeProvider` and use its associated types? Or something else?

Reply:
I'm open to discussion on this, I don't love either solution. My thinking is that we already have a bunch of public interfaces parametrized with the runtime provider, and it feels simpler as a public API to extend that here compared to what's so far functions more like an implementation detail of the runtime provider implementation.

Like, I think the "runtime provider" is a somewhat reasonable straightforward concept to have as part of the public API - we're effectively abstracting over the async runtime. Splitting out this trait -- so far mostly used in internals as part of the runtime provider abstraction -- feels like it makes it a more salient part of the public API, which I worry might make the API harder to understand/use.

Reply:
> Splitting out this trait -- so far mostly used in internals as part of the runtime provider abstraction -- feels like it makes it a more salient part of the public API, which I worry might make the API harder to understand/use.

That's a reasonable argument. I don't feel super strongly. Mostly I wanted to understand the motivations one way or the other.

Reply:
This is turning out to be rather messy in the case of `Client`, which also implements `DnsHandle`. Adding generic type parameters in the right places, so they can flow to the associated type, is rather involved, and will require either a turbofish or another parameter on `Client::new()`, for example. The client builders and constructors could use some significant clean-up, especially to replace manually-implemented futures with async functions.

Reply:
I _feel_ like the right way would be to return the normal `Error` and then match on Error::Io later
https://github.com/hickory-dns/hickory-dns/blob/3d18499eeadbeb5cd0e3a8386a5eedc0a1ac08d7/crates/proto/src/error.rs#L242

Reply:
I have rebased this on top of the `DnsHandle` changes.